Navigation updates (CSS for desktop and the start of mobile)

garrett opened the pull request:
This isn't done — the changes I've made for desktop are getting there. As I accidentally looked at things in mobile responsive mode and saw how horribly they were broken (unusably broken), it prompted me to try to tackle the mobile UI as well, so there's a WIP in a separate commit in here as well.

This PR will need to be rebased on master and the navigation still needs some work. The mobile part needs even more work.

I'll add a new mobile nav mockup I made today in the comments.

garrett commented:
# Mockup

I'm moving the WIP code in this branch toward this concept:

![nav-mobile-bottom](https://user-images.githubusercontent.com/10246/29176059-0981380e-7deb-11e7-92f0-9c8a25485219.png)

The basic changes needed for it are already implemented, but it still needs a lot of work.

A quick write-up of the thought process behind it:

## Bottom nav

When people use phones, their hands are typically at the bottom of the screen. Moving the navigation closer to people's hands means they can jump around and navigate Cockpit more easily. 

## Interacting with the nav

Tapping on the icons on the bottom switches the section and expands the menu. If Cockpit is already at that section, the menu is toggled. The menu may be in the shell or it might be in the iframe. If in the iframe, it's positioned absolutely.

If there are more than 4 (or possibly 5?) items, then it should become horizontally scrolly (and visually indicate this). In most cases, people will hopefully not have so many modules in their installations.

## Position of the menu

The menu doesn't go all the way to the top or to the right. It stops on the right at the same width of the three-line menu icon. Tapping the non-menu areas will dismiss the menu. Tapping the sub-navigation (systems, services, containers, logs, etc.) will select that section and also dismiss the menu.

## Non-module three-line "hamburger" menu

The lock and user menu are shoved under the hamburger menu on the bottom right. (The rest of the icons are always left-aligned, and may stretch or squeeze a bit using min/max width depending on how many items there are.)

## What do we do with the heading?

We _could_ have the heading at the top and scroll off with the content — but I think it's fine to omit it. The sign in page has branding and machine that's currently active and I think it's probably fine to just move it under the three-line "hamburger" menu on the right anyway.

## Main goals

Anyway, the main goal is to free up as much space as possible to make the content be the star. The navigation should always be available though, so I hope I struck the right balance with this. What do you think?

## Implementation details

I'm pretty confident we can do almost all of this just in CSS. There will need to be a little bit of JavaScript to handle the events just right, I'm sure.

As far as 3rd party apps are concerned, I think we can make this all work within an iframe too. If they don't do extra work to make their app mobile-friendly, then the worst case scenario is that if they have a menu it will be left-aligned and static (not ideal, but not the end of the world) or they might not even have a menu anyway (like dashboard).

## In closing

Even in an ideal situation, it will probably always be a little better to access Cockpit in a desktop web browser — but making Cockpit also available to do perform from a mobile device will be a "killer feature" in the future — and navigation is the first step to allow us to do this.

garrett commented:
In case you're wondering what the current status is right now, it looks like this on mobile (this is a direct screenshot from within Firefox):

---
![cockpit-responive-screenshot](https://user-images.githubusercontent.com/10246/29176901-825c144a-7ded-11e7-9a28-7e752db9da2a.png)
---

...there's still a bit more to go. :wink: But I do have the layout shifting around (similar to what I mocked up) for mobile already.

BTW: That menu height is currently hardcoded — but it is scrolly, works, and can access any every of Cockpit already.

petervo commented:
I rebased and pushed.

I'm not sure about removing the recalculate layout part. For me in FF the page no longer takes up all the vertical space now. before that function was what was setting the frame height making it fill the page.

garrett commented:
Oh that's odd. It's working for me in Firefox. Yeah, it needs to be tested further in different scenarios and fixed where it breaks. If we can make it work under every circumstance without the JS, that'd be great. (If it does need to go back in, then it shouldn't do as much as what it does currently.)

garrett commented:
(I might've also broken something when testing & adjusting things in Chrome and also on mobile in Firefox too.)

garrett commented:
_Huge_ CSS changes:

- put desktop-specific CSS in a media query
- reworked mobile CSS using grids

(FYI: I moved the older WIP flexbox version to https://github.com/garrett/cockpit/tree/navigation-mobile-flexbox and changed this branch to the grid version)

# Screenshots

All screenshots are taken with my Nexus 6P running Firefox Nightly. This also works in Chrome and Chromium-based browsers, such as Brave. It should work similarly on an iPhone (but I don't have an iOS device to test).

## No menus

I faked this one, as I have the subnav as a fixed menu right now — it needs some JS-based toggle interactions hooked up to work properly, and the height should be adjusted once that's done... (so for now, it's just a `display: none;` to hide it for some screenshots):
![screenshot_20170921-154451](https://user-images.githubusercontent.com/10246/30705411-adc74018-9ef5-11e7-9426-2e1d95e4813c.png)

## Dashboard menu

This shows up when you would switch to a dashboard. Selecting an item dismisses it. Tapping the icon again when you're in the dashboard shows the menu once more. It would be nice to have some animation when it is shown (I think I can handle that animation in pure CSS):
![screenshot_20170921-154312](https://user-images.githubusercontent.com/10246/30705510-ee075708-9ef5-11e7-8dee-b68c080a0a46.png)

## Nav overflow ("hamburger") menu

Highlighting on the navbar is wrong, but it should be pretty straightforward to fix.

The menu should be expanded by default (right now you have to tap on your name to see more items; they should be always visible when the menu is seen).
![screenshot_20170921-154421](https://user-images.githubusercontent.com/10246/30705584-200967dc-9ef6-11e7-84c4-6a934b91183c.png)

## Dialog

Most everything works as expected in mobile mode, including this language dialog (triggered by the nav overflow menu):
![screenshot_20170921-154440](https://user-images.githubusercontent.com/10246/30705672-607095d4-9ef6-11e7-9b71-cafd2bbd697a.png)

# Summary

This is actual working code (_only_ CSS changes!) to make it look like the mockup. There are a few minor things that need fixing (interaction when the menus are called), but it should be getting pretty close overall.

(Note: This pass does _not_ include subnavigation in iframes... yet. That will be a little trickier, but should be possible still. We may have to rely on `position: fixed;`, as well as communicate to the document in the iframe to make things act similarly.)

garrett commented:
For reference, here's the current state of the navigation in desktop mode (nothing has changed in this commit):

![screenshot-2017-9-21 system - sunny](https://user-images.githubusercontent.com/10246/30705992-60327884-9ef7-11e7-9630-2c9f727c12ef.png)

garrett commented:
Note: I want to add more comments to the CSS, as using CSS grids is a very new way of positioning in CSS and I did some special stuff to make it work as intended (100% standards compliant without hacks), so it might not be that easy to follow along unless I explain it a bit better in the code.

(**TL;DR**: I'll add more comments next week.)

petervo commented:
@garrett I rebased. Both me and andreas are having issues with frame size. see https://pasteboard.co/GLztWtr.png. Additionally in mobile mode, only the machine link shows up. No dashboards show at all for me.

martinpitt commented:
In firefox, in "Nexus 7" (small tablet) format I get exactly the same issue as @petervo and @andreasn - half of the left nav bar and ¾ of the main contents are just white and cut off.

On a phone format, the popup menu is open by default and can never be closed:

![phone](https://user-images.githubusercontent.com/200109/30804264-08809442-a1ed-11e7-9a97-472499fd2e30.png)

And I also can't see any dashboards.

I confirmed this behaviour on my actual phone (Sony Xperia X Compact), so it's not just the Firefox emulation.

andreasn commented:
I've got things working by setting .`flex-sidebar { height: 100%;}` and `.container-frame {height: 100%;}`

andreasn commented:
When I hover things in the sidebar, I get the text decoration underline and the pointer, but no indication of the size of the clickable area, so I think a background color on hover would help here.

garrett commented:
After @petervo helped me un-mess-up my system yesterday (huge thanks!) I worked on fixing up the desktop navigation and started the mobile nav. Today, I mainly reworked the mobile nav and tweaked both desktop and mobile.

It should work much better for you both as of today.

The TODO list is basically the same as before, with the subnav menu needing a way to toggle it and the overall menu needing to be expanded fully when it is shown.

In addition to the nav work, I also adjusted the host switcher to more match the mockups.

I checked to make sure `single-nav` mode still works, and it does in desktop... in mobile, it still shows the bar, as that's how navigation works on mobile (unless we redesign it).

garrett commented:
## Desktop

### Dashboard

![screenshot-2017-9-26 dashboard - sunny](https://user-images.githubusercontent.com/10246/30863125-85eacc48-a2d0-11e7-9b5b-0926b3d24e29.png)

### Remote server + host switcher

![screenshot-2017-9-26 storage - dew](https://user-images.githubusercontent.com/10246/30863124-85e22c64-a2d0-11e7-8c39-1f949df761d2.png)

### Single-nav

![screenshot-2017-9-26 storage - sunny](https://user-images.githubusercontent.com/10246/30863122-85cd2198-a2d0-11e7-8699-675c33fe3210.png)

## Mobile

### Sub-nav (which is currently hard-fixed; we need to make it toggleable)

![screenshot_20170926-153345](https://user-images.githubusercontent.com/10246/30863626-d19153dc-a2d1-11e7-8e16-f63a9884444b.png)

### Dashboard

![screenshot_20170926-153352](https://user-images.githubusercontent.com/10246/30863623-d1210f1e-a2d1-11e7-9122-410c6a87fddd.png)

### Registry 404 (note the slidy nav icons to switch dashboards!)

![screenshot_20170926-153404](https://user-images.githubusercontent.com/10246/30863624-d1272430-a2d1-11e7-8b9b-2bf5262e2aff.png)

### Site menu

![screenshot_20170926-153414](https://user-images.githubusercontent.com/10246/30863625-d131df10-a2d1-11e7-8bb7-f40d87c81340.png)

garrett commented:
Note: If you're testing mobile in responsive mode, the sliding nav icons at the bottom might not work on your desktop browser. They work on an actual mobile device and in Chrome's responsive mode (which also isn't quite like testing on a real device, but for other reasons — especially when it comes to the interpretation of 100vh, which is frustratingly slightly off).

garrett commented:
Screenshots of nav tweaks:

## Subnav fade prominence

![screenshot_20170926-175226](https://user-images.githubusercontent.com/10246/30870234-a4b8b564-a2e3-11e7-856a-ab3930857038.png)

## Line at the bottom & un-highlighted unselected icons

![screenshot_20170926-175243](https://user-images.githubusercontent.com/10246/30870324-deebeda0-a2e3-11e7-8f36-44520da11072.png)

garrett commented:
Current TODO:

- [ ] Toggle subnav menu
- [ ] Dismiss subnav menu when item (in the scrollable area) is chosen
- [ ] Increase height of the subnav (once it's dismissable)
- [ ] Have shell menu fully expanded (items under username) when it is activated

petervo commented:
Pushed a commit that might help with the host navigation toggle for mobile.

There is an issue where the top nav shows empty on machine troubleshooting pages.

In chrome for me the search icon's alignment is off.

I also need to hookup the machine search js.

garrett commented:
## Safari incompatibility 
Issue @ https://github.com/cockpit-project/cockpit/issues/3640 — I guess this isn't due to the nav, at least. Also, I can recreate the issue with GNOME-Web (aka: Epiphany), which I guess really is the Safari of the GNOME desktop. :wink: Having Epiphany also affected might help us better debug things? 

The best most up-to-date Mac I have here is from mid-2009, which hasn't gotten the past couple of macOS updates. So it might be representative or not. I think Epiphany is more likely to be aligned with what's in recent macOS, due to sharing the same rendering engine, WebKit.

## Minor style issues

I'll work on these along with the other mentioned style tweaks and get them merged in soon. I think it shouldn't really affect the work that @petervo is doing.

garrett commented:
- I fixed the curtains issue so that it's now properly aligned in the content area and nav still shows up.

- I've also hooked up the CSS necessary for the hosts menu to show up properly with the JS @petervo added. The interactions still need refinement; I'll discuss with @andreasn to figure out what might feel best. (I have ideas, but I want to confirm.)

Now that I've landed these two above patches, Cockpit should start to work pretty decently on Mobile... at least for the hosts and dashboard areas. There are obvious things that need fixing for mobile in various parts, but we can now start to see and interact with everything from a phone. _This is exciting!_

(I'm still working on various style tweaks, but I'll land them soon. I stashed them out of the way to fix these high-priority issues first.)

petervo commented:
We need to limit the height of the host navigation menu (system, docker etc...) in mobile mode so that the top is not off screen.

petervo commented:
Overflow of dashboards in mobile needs some work.  Both long machine names and lots of dashboards can make some links not accessible.

garrett commented:
@petervo: 
1. The host navigation menu is height-limited in mobile. I did change how it was calculated recently, however.
2. When icons overflow, you're supposed to slide them over to the side. It's a common pattern in mobile apps and websites these days. (That said, I should re-add label truncation... but the issue is related to "Image Registry" which is pretty long and would be truncated by default.)

garrett commented:
- Fixed subnav host menu for Chrome on Mobile by explicitly setting a calculated height for all browsers (including Chrome). It's annoying how `100vh` isn't `100vh` on Chrome mobile. 😞

petervo commented:
> When icons overflow, you're supposed to slide them over to the side. It's a common pattern in mobile apps and websites these days. (That said, I should re-add label truncation... but the issue is related to "Image Registry" which is pretty long and would be truncated by default.)

No it was caused by a long hostname for me.

garrett commented:
Long hostnames are now truncated in the nav icons

garrett commented:
I think I addressed all @petervo's & @andreasn's concerns.

@petervo: One additional thing that I see missing is having the current host selected in the host switcher.

Also, the layout is a bit odd on initial load on Chrome on Android at the moment. (It's probably due to the menu animation? I'll look to see if I can fix it.)

garrett commented:
Ok, I believe I fixed the initial layout issue with Chrome on Android by changing the way the menu animates.

petervo commented:
In IE, the host submenu, host-nav doesn't scroll.

petervo commented:
@garrett we are still having some iframe height problems in the tests, where tests fail because not enough of the frame is visible.

garrett commented:
At least here, in testing, I have fixed IE's buggy layout issue in 5aae421069b43f3f81eca28f12a030cc7c29764e by adding flex-basis as 100vh (that's 100% of the vertical height as a starting point for size calculation).

There's a bit of a possibility that it might fix old Chrome (47.0.2526.80 from Dec 2015, used in our Selenium tests) too, but we'll have to wait and see what happens with the test. (Current Chrome acts properly.)

garrett commented:
With a little bit of searching around, I found a way to test Chromium of the exact same version as what's used in our tests. There was an update for openSUSE 13.2 that included the same exact version that still exists. It's at http://ftp.gwdg.de/pub/opensuse/discontinued/update/13.2/x86_64/chromium-47.0.2526.80-61.1.x86_64.rpm — so I'm downloading and installing openSUSE 13.2 (available by torrent still) and that RPM.

This was from the era where Chromium wasn't shipped by Fedora. Ubuntu doesn't have the archives easily laying around, and it doesn't seem that specific version was available in Ubuntu anyway. And the official built snapshot doesn't easily/properly run on a modern Fedora. So old openSUSE in a VM it is! Fun.

garrett commented:
I think I found a balance to make even old Chrome work, while not breaking other browsers (including IE) in commit 1e1438a: Absolute positioning for desktop views. It apparently works fine in Firefox, Chrome, and IE11.

andreasn commented:
So maybe this is included in the "Improve host switcher style (including hover states)", or not.
But there seems to be an issue with the width of the input in the host switcher.
![screenshot from 2017-10-05 13-36-56](https://user-images.githubusercontent.com/209895/31225602-1e725608-a9d3-11e7-9b9e-236a7feaa928.png)
Looks like it's due to `.has-feedback .form-control {padding-right: 32.5px;}` in forms.less

garrett commented:
The switcher style needs active host highlighting and the triangle.

What's the issue with the width? Is it the padding on the right that should only be on the left? (Just judging from your screenshot.)

garrett commented:
Ah, right, yes. Sorry; brain still halfway out to lunch. :wink: Time for the second coffee ☕️.

garrett commented:
Commit b3d1f2a is a shot in the dark in attempt to (mostly) placate Phantom.js tests with respect to content areas.

garrett commented:
(Using `git revert` for now until I figure out what's a useful removal, then we can git rebase to drop the changes out.)

garrett commented:
I rebased the changes I had locally on top of the changes from yesterday and undid the reverts, so all the attempted work-around commits are now in the branch again.

petervo commented:
@garrett so it looked like jquery was doing the wrong thing when  as forcing it to the wrong thing when the css was falling back to it was falling back to the webkit fallbacks. That and forcing opacity seems to work. 

There is still a overflow problem in that mode. Can you take a look at that as well as squash the commits down and remove anything you don't think is needed. it's getting pretty unwieldy.

There are also phantomjs crashes on debian, i'll look into that tomorrow.

garrett commented:
@petervo: OK! I've squashed everything in a rebase and also added some possible fixes.

In case we need to look at what was done before during development, I have the previous state of the branch @ https://github.com/garrett/cockpit/tree/navigation-mobile-reference

garrett commented on pkg/shell/indexes.js:
```diff
@@ -458,6 +458,8 @@ var phantom_checkpoint = phantom_checkpoint || function () { };
                         .css("color", machine.color || ""));
                 return $("<li role='presentation'>")
+                    // FIXME: This isn't updated on switch
+                    .toggleClass("active", machine.key == machines.lookup()['key'])
```
I'm pretty sure one or both of the following is true:

1. This isn't the proper way to check the current machine versus the one in the list
2. This probably is called to create the list once, not every time (didn't check)

I left this in as a FIXME as it adds the 'active' class and shows the current machine on page load at least. I figure you could do a more proper fix and we can squash it together later.

petervo commented:
@andreasn @garrett I think this is ready for another review from the design side.

The code needs a proper review as well.

garrett commented:
@petervo: Update on the host switcher: By default, it doesn't show a selected host here. However, when I switch to another machine, that one is highlighted. Also, curiously, when I switch back, the original machine is highlighted as well! (It's just not highlighted by default.)

petervo commented:
@garrett pushed an update

petervo commented:
@martinpitt or @larskarlitski could one of you review this as well?

garrett commented:
So I still feel the CSS uses too many IDs (it should uses classes in probably every place it uses IDs right now); this needs to also be fixed in the HTML (by adding the classes to the various places) and it _might_ affect the priority, as IDs are more important in the cascade (specificity rules), so weird things might happen and various measures might need to be taken.

Also, I'd like for us to rename the `.flex-`* classes to something else, since we're not using flexbox in mobile mode. What, though? struct- for structure? or something else? Suggestions?

It's ideal if we can get this changed before merge, but we could also have another PR to tidy this up.

andreasn commented:
Last commit (https://github.com/cockpit-project/cockpit/pull/7482/commits/005b3b57a8ba02cbc1505972a6ba08292e33ad65) messed it up a bit on my cell phone (using Chrome)
![photo_2017-10-26_18-30-23](https://user-images.githubusercontent.com/209895/32065114-17cfb48a-ba7c-11e7-9508-42d34da3f3f3.jpg)

andreasn commented:
I tested briefly on the latest version of Edge (16).
It seems to have some issues in the top area. No branding, and there is a lack of bottom padding.
![index](https://user-images.githubusercontent.com/209895/32096478-e2df8da0-bb06-11e7-833e-0ac0d554385b.png)

larskarlitski commented:
Awesome work! Looks and works like advertised to me except on mobile Safari. The navbar scrolls with the output :(

petervo commented:
@andreasn did you try master on edge? Is that actually a regression?

andreasn commented:
> did you try master on edge? Is the branding missing actually a regression?

Tried the version that came with Fedora 26 now, and it doesn't show the branding there either, so not a regression. Filed it as #8002

martinpitt commented on bots/naughty/rhel-7-4/7668-machines-jump:
```diff
@@ -1,2 +1,2 @@
-  File "test/verify/check-dashboard", line 214, in testBasic
+  File "test/verify/check-dashboard", line 218, in testBasic
```
Better replace the line number with a `*`.

martinpitt commented on test/verify/check-pages:
```diff
@@ -114,8 +114,15 @@ WantedBy=default.target
+            else:
+                b.wait_not_present(".flex-body.single-nav")
+            b.wait_present("#host-nav")
+        elif "fedora-atomic" in m.image:
```
This is a dead branch now, as any atomic image already satisfies `m.image != "rhel-7-4"`.

martinpitt commented:
Semaphore tests time out consistently - can you please rebase to current master, to ensure this has the recent fixes for semaphore?

garrett commented:
I rebased; semaphore passed now.

Most tests pass, but a couple (fedora-27 & ubuntu-1604 right now) are still failing (with unrelated issues from what I could tell — timeouts, it seems?)

martinpitt commented:
The other two failures on fedora-27 are flakes (they succeeded in the previous runs), but this one seems to be very persistent (5 failures in a row): https://fedorapeople.org/groups/cockpit/logs/pull-7482-20171108-165656-4b8a37d9-verify-fedora-27/log.html#114

petervo commented:
@garrett, can you squash this down a bit before we start fiddling with trying to get the tests to pass?

petervo commented:
@garrett can confirm that the sidebar is hidden for kubernetes. This is a regression since https://github.com/cockpit-project/cockpit/pull/7482/commits/e7d3a08d03e09685a7a23e4c1de562725965f8b1

To test, run
```
mkdir -p ~/.kube
ln -s /path/to/cockpit/bots/images/files/openshift.kubeconfig ~/.kube/config
./test/vm-run openshift
```

martinpitt commented:
For some reason the rhel* tests weren't triggered automatically, I did that manually now.

petervo commented:
@garrett, looks like the original fix for this was lost somewhere along the way. I pushed another one, but would be good for you to review.

petervo commented:
@martinpitt, debugging elsewhere, see: https://fedorapeople.org/groups/cockpit/logs/pull-8043-20171109-024747-99c97f93-verify-debian-testing/log.html#116

garrett commented:
@petervo: I looked at the CSS and ran the openshift/kube stuff again and it looks good as far as I can tell.

garrett commented:
- merged master
- rebased on top of master
- squashed everything feature related into 1 commit
- added design URLs to commit text
- kept systemd & bots as separate commits

martinpitt commented:
The tests prior to rebasing were green, this is now spoiled by the unrelated IPA regression and some flakes. Good enough for merging now IMHO. Thanks, great work here!

martinpitt commented:
Argh, forgot to add a Closes #, mea culpa.

garrett commented:
Here are a few screenshots (I faked some of the details so it's not my own machine's IDs)

## Desktop

![c104f094-37b9-43b5-a3a4-3647e3301b2a](https://user-images.githubusercontent.com/10246/32792280-4931bdf2-c963-11e7-9bb1-0762f5e2e981.png)

## Mobile

| Menu closed | Menu open |
|---|---|
| ![f2aedf90-9593-46f3-958e-1f7ca42202eb](https://user-images.githubusercontent.com/10246/32792306-56a3be5e-c963-11e7-9425-43c381cf79cb.png) | ![8a4d5e7e-bbe5-4d98-825a-8689d68a8701](https://user-images.githubusercontent.com/10246/32792314-5c057a40-c963-11e7-8e34-4fa67516d59e.png) |

---

I noticed that the implementation doesn't align the host name with the side navigation in the desktop, as in the original mockups. Oops. (You can really see it the highlight on "System" overview.) I guess I need to fix this minor thing and send another PR soon.
